DRAFT: Feature/updated models

nmenezes0 opened the pull request:
## Context

<!-- Why are you making this change? What might surprise someone about it? -->
Adding new models as discussed previously - to reflect new pipeline and theme generation.

Have added a "minimal" version of models i.e. just what we need to fit with the current dashboard pages.
There are some other fields/models, but they can be added as and when needed.

**Warning** - merging this PR will delete the existing consultation data in the database on prod. (Will not delete users.)

## Changes proposed in this pull request

<!-- If there are UI changes, please include Before and After screenshots. -->
* Add new models - except for the `User` model, 
....

## Guidance to review

<!-- How could someone else check this work? Which parts do you want more feedback on? -->

## Link to Trello ticket

https://trello.com/c/PHi7K23V/27-django-data-models-mvp-v1

Though doesn't add all of the models

## Things to check

- [ ] I have added any new ENV vars in all deployed environments and updated the `.env.example` and `.env.test` files in the repo

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+    # TODO - add metadata e.g. langfuse_id
+    # Note, the execution run will be run on responses to a particular
+    # question part - but this will be stored in the correspondong framework/mapping etc.
+    history = HistoricalRecords()
```
In which cases would these get edited? Do we need historical records here?

nmenezes0 replied:
Good challenge, we thought they might when we had SQL table convos but if I'm aiming to be minimal, the `HistoricalRecords` should come out and we can re-add if/when needed.

nmenezes0 replied:
`HistoricalRecords` have now been removed for the `ExecutionRun` model (https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92).

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
     consultation = models.ForeignKey(Consultation, on_delete=models.CASCADE)
-    name = models.TextField()
-    slug = models.CharField(null=False, max_length=256)
+    order = models.IntegerField(null=True)
```
Couple of questions: should these be unique for a given consultation? And do we want them to be nullable?

rachaelcodes replied:
followup question, do we want to overwrite `ordering` so that it's by order rather than `created_at`?

nmenezes0 replied:
Actually yes, they should be unique and non-nullable. Now I think about it more, why would you want your questions to be in a random order?

Didn't think of that - overwriting ordering sounds good.

nmenezes0 replied:
This field is now called `number` (raised by Duncan elsewhere), and we now have constraints so they are unique up to consultation. 

And I've rewritten the ordering to order by number.

nmenezes0 replied:
https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92 - change `order` to `number`, make unique and non-nullable, order on this field (for `Question` and `QuestionPart`).

rachaelcodes replied:
Great! Thanks for putting all the related changes in one commit (much easier to follow).

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+    text = models.TextField()
+    type = models.CharField(max_length=16, choices=QuestionType.choices)
+    options = models.JSONField(default=list)
+    order = models.IntegerField(null=True)
```
Same questions as for `class Question`

nmenezes0 replied:
And same response - what you said above seems sensible.

nmenezes0 replied:
Changed as described above for `Question`.

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+    execution_run = models.ForeignKey(ExecutionRun, on_delete=models.CASCADE, null=True)
+    question_part = models.ForeignKey(QuestionPart, on_delete=models.CASCADE)
+    # When Framework is created - record reason it was changed & user that created it
+    change_reason = models.CharField(max_length=256)
```
Do we know from UR whether we need this yet? Or can we hold off on it for now?

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+    # change the theme.
+    answer = models.ForeignKey(Answer, on_delete=models.CASCADE)
+    theme = models.ForeignKey(Theme, on_delete=models.CASCADE)
+    reason = models.TextField()
```
Is this the change reason? Or the reason for the mapping decision?

nmenezes0 replied:
This is the reason for the mapping decision and is an output of the pipeline, but actually, this might have been removed. (And if we are including we need another field for `quote`.)

@alexmoore-iai - are we outputting the `quote` and `reason` in the theme mapping in the current version of the pipeline?

alexmoore-iai replied:
So current outputs from the theme mapping are labels, reasons and stances (the last of these only went in last week). Quote isn't included for now, but may have to go back in depending on UR feedback

https://github.com/i-dot-ai/consult-ai/blob/main/consult_ai/consult_ai/tasks/prompts/theme_mapping.txt

nmenezes0 replied:
I've left the `reason` in and created a ticket to add the other fields (need to check what they are/options etc).

duncanjbrown commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+                ModelClass.objects.filter(slug=generated_slug).exclude(pk=self.pk).exists()
+            )
+        else:
+            slug_exists = ModelClass.objects.filter(slug=generated_slug).exists()
```
What does this if statement get us that `UniqueConstraint` wouldn't?

nmenezes0 replied:
What the `save()` does is:
* Generate slug from text
* If slug is unique - save that and we're done
* If slug is not unique - make it unique by adding a timestamp at the end (though will happen infrequently)
`UniqueConstraint` we would have to wait til we've saved to find out the slug was invalid.

duncanjbrown commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+        if not generated_slug:
+            generated_slug = timestamp
+        elif slug_exists:
+            generated_slug = f"{generated_slug}-{timestamp}"
```
what do we need from timestamps in slugs?

nmenezes0 replied:
The timestamp is for uniqueness, on the off chance that two questions have the same slug (unlikely in general, but as I found out last time, will def happen a lot in testing when we're uploading the same consultation loads of times).

nmenezes0 replied:
I've also just realised the question slug only needs to be unique up to consultation, not unique overall.

duncanjbrown commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+        pass
 
-    def add_llm_summarisation_detail(self):
+    def amend_framework(self, user: User, change_reason: str) -> "Framework":
```
This method name is a bit misleading I think as it doesn't do amending — it creates a descendant?

nmenezes0 replied:
I remember struggling with the naming! `create_descendant_framework`?

duncanjbrown replied:
better!

nmenezes0 replied:
Now renamed: https://github.com/i-dot-ai/consult/blob/77148101c0e62e4d5cdf29d2cbf8b0401aa2c830/consultation_analyser/consultations/models.py#L192
Also similar renaming for themes.

duncanjbrown commented on tests/unit/models/test_framework.py:
```diff
@@ -0,0 +1,87 @@
+from consultation_analyser.consultations import models
+
+
+def set_up_frameworks():
```
A bit of a nit, but this is edging on "mystery guests" in the tests, where to read and understand the tests we just have to know that the different frameworks this method returns have particular relationships with each other. Giving them more explicit names, and maybe breaking them into fixtures, could help with this? that said, I'm not sure how whether pytest fixtures can depend on each other, so it might be that this is the pragmatic approach!

nmenezes0 replied:
Yes, will think about how to do this better. With some of the changes we discussed on Wed, have to change the factories anyway, so will be changing the tests.

nmenezes0 replied:
Have totally rewritten these tests with subsequent changes: https://github.com/i-dot-ai/consult/blob/feature/updated-models/tests/unit/models/test_framework.py - any thoughts on how to improve?

duncanjbrown commented on tests/unit/models/test_theme_mapping.py:
```diff
@@ -0,0 +1,68 @@
+    first_execution_run = ExecutionRunFactory(type=models.ExecutionRun.TaskType.THEME_MAPPING)
+    ThemeMappingFactory(answer=answer, theme=theme1, execution_run=first_execution_run)
+
+    # Second theme mapping execution run
```
We're almost writing app code here! It would be nice to be able to hit a "map_themes" method or something and have it actually run the pipeline (a fake pipeline in tests, of course) and write the appropriate records?

nmenezes0 replied:
I decided not to do this in the interests of getting this PR over soon... (but happy to be overruled). Thought it best to do this when we add the pipeline anyway (or if it turns out to be useful in other tests).

nmenezes0 commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
+    question = models.ForeignKey(Question, on_delete=models.CASCADE)
+    text = models.TextField()
+    type = models.CharField(max_length=16, choices=QuestionType.choices)
+    options = models.JSONField(default=list)
```
Actually yes, it probably should be null for free text instead of empty list. Then I think this affects the factories above that you commented on.

nmenezes0 replied:
Yes, now changed to default to null (which it will be when it's free-text) (https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92).

rachaelcodes commented on consultation_analyser/consultations/views/answers.py:
```diff
@@ -0,0 +1,57 @@
+        slug=question_slug,
+        consultation=consultation,
+    )
+    # if processing_run:
```
Do we need to keep this? Have we got rid of processing runs?

nmenezes0 replied:
Good spot - yes we have. Will remove.

nmenezes0 replied:
Now deleted.

rachaelcodes commented on consultation_analyser/consultations/views/answers.py:
```diff
@@ -0,0 +1,57 @@
+        free_text_answers = models.Answer.objects.filter(question_part=free_text_question_part)
+        total_responses = free_text_answers.count()
+    else:
+        free_text_answers = models.Answer.objects.none()
```
TIL. Interesting

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,237 @@ class Meta:
-    counts: dict[str, int]
-    total_responses: int
-    has_multiple_selections: bool
+class QuestionPart(UUIDPrimaryKeyModel, TimeStampedModel):
```
Given that there are a few places where we're assuming there's only one free text type per question, do we want to include some model validation to confirm this?

nmenezes0 replied:
Yes - that's a good idea.

nmenezes0 replied:
Uniqueness constraint (at most one free-text part per question) added: https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92.

rachaelcodes commented on tests/integration/test_consultation_page.py:
```diff
@@ -1,20 +1,23 @@
     consultation_slug = consultation.slug
-    question = consultation.section_set.first().question_set.first()
+    question = Question.objects.filter(consultation=consultation).first()
     homepage = django_app.get(f"/consultations/{consultation_slug}/")
```
A bit picky, so I apologise, but is `homepage` the right variable name here?

nmenezes0 replied:
probably not - consultations or similar I think

nmenezes0 replied:
Now updated https://github.com/i-dot-ai/consult/blob/77148101c0e62e4d5cdf29d2cbf8b0401aa2c830/tests/integration/test_consultation_page.py#L19.

rachaelcodes commented on tests/unit/models/test_theme.py:
```diff
@@ -0,0 +1,48 @@
+    theme_x = models.Theme.objects.get(name="X", framework=framework_1)
+    models.Theme.objects.get(name="Y", framework=framework_1)
+
+    with pytest.raises(ValueError):
```
Is it worth asserting the error message? If nothing else, it might make it clearer what the problem is when reading the test.

nmenezes0 replied:
Updated: https://github.com/i-dot-ai/consult/blob/77148101c0e62e4d5cdf29d2cbf8b0401aa2c830/tests/unit/models/test_theme.py#L39

rachaelcodes commented on tests/unit/models/test_theme_mapping.py:
```diff
@@ -0,0 +1,68 @@
+    question_part = QuestionPartFactory()
+    answer = AnswerFactory(question_part=question_part)
+
+    # First theme mapping execution run
```
These test explanation comments are very helpful :)

rachaelcodes commented on tests/unit/models/test_theme_mapping.py:
```diff
@@ -0,0 +1,68 @@
+
+
+@pytest.mark.django_db
+def test_get_history_of_changes_to_answer():
```
Is this the right name for this test? It made me think it would be reading the change history all at once.

nmenezes0 replied:
Does this make more sense? https://github.com/i-dot-ai/consult/blob/77148101c0e62e4d5cdf29d2cbf8b0401aa2c830/tests/unit/models/test_theme_mapping.py#L48

rachaelcodes replied:
yes, thanks!

nmenezes0 commented on consultation_analyser/consultations/dummy_data.py:
```diff
@@ -1,59 +1,117 @@
+    for question_data in questions_data:
+        question = QuestionFactory(
+            text=question_data["question_text"],
+            order=question_data["order"],
```
This has now been changed to be `number` for both the `Question` and `QuestionPart` (https://github.com/i-dot-ai/consult/pull/405/commits/517f14391a6c9008a953ce9152e36ae3e20adb92).

nmenezes0 commented on consultation_analyser/consultations/dummy_data.py:
```diff
@@ -1,59 +1,117 @@
+                question=question,
+                text=part["text"],
+                type=question_part_type,
+                options=part.get("options", []),
```
Now we assign null for free text questions - the dummy data generation has been written to use updated factories: https://github.com/i-dot-ai/consult/blob/feature/updated-models/consultation_analyser/consultations/dummy_data.py

nmenezes0 commented:
I have addressed all comments from the code review.

Additionally, I have updated the `save()` methods from `Theme` and `Framework` as discussed in our meeting last week: https://github.com/i-dot-ai/consult/blob/feature/updated-models/consultation_analyser/consultations/models.py. **Please can you look over these changes.**

Additionally I have added some constraints to the `number` field (formerly the `order` field) - so you'll need to drop your DB locally. I don't think this will cause problems in the deployed environments (dev/preprod/prod) as there isn't any existing data there (for the new models). To replicate, on `main` generate some dummy data, then switch to this branch and run migrations and check nothing is broken. **Is this ok, do you agree that we don't need to worry about these changes?**

rachaelcodes reviewed (approved):
The changes all look good to me - I guess you might want to see what Duncan has to say too.

rachaelcodes commented on consultation_analyser/consultations/models.py:
```diff
@@ -40,301 +41,277 @@ class Meta:
     consultation = models.ForeignKey(Consultation, on_delete=models.CASCADE)
-    name = models.TextField()
-    slug = models.CharField(null=False, max_length=256)
+    number = models.IntegerField(null=False, default=0)
```
A question - just noticed this (sorry it's after reviewing) - does integer numbering work with the looser number typing in `consult-ai`: https://github.com/i-dot-ai/consult-ai/blob/main/consult_ai/consult_ai/io/models.py#L66 (maybe it does, but then we might be assuming we get given the data in the right order from them?)

nmenezes0 replied:
As discussed on Slack - we'll use `number` to denote the order and keep as an integer.

If needed in future we can incorporate a display name or similar.

nmenezes0 commented:
Will mark other comments as resolved as per our convo @duncanjbrown and @rachaelcodes. Will make the final change summarised below.

On slugs and uniqueness:
* Question - assume question always has text, for uniqueness, append question number if needed.
* Consultation - append with an integer (incrementing) if the slug already exists.

nmenezes0 commented:
I have resolved all conversations apart from discussion around slug fields (as discussed yesterday, that was the only outstanding thing).

I have now made changes to the slug fields, so we add an integer if they are not unique (though we don't expect this to happen often). 

I have also renamed `text` field in `Consultation` model to be `title`.

nmenezes0 commented on consultation_analyser/consultations/models.py:
```diff
@@ -70,21 +70,63 @@ class Meta:
+        cropped_text = self.text[:cropped_length]
+        generated_slug = slugify(cropped_text)
+        if self.pk:
+            slug_exists = Question.objects.filter(slug=generated_slug).exclude(pk=self.pk).exists()
```
Doh, yes!

nmenezes0 commented on consultation_analyser/consultations/models.py:
```diff
@@ -70,21 +70,63 @@ class Meta:
+                return False
+            return Consultation.objects.filter(slug=slug).exists()
+
+        cropped_length = 220
```
I think that was a legacy from the timestamp - but it makes sense to crop later.
